### PR TITLE
test: Add testWithFlameGame and testWithGame helper functions

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

Add function `withFlameGame()` that can supply a prepared `game` instance to the test body. 

The main advantage over the `flameGame.test` is that it doesn't confuse Android Studio over whether something is a test or not. In particular, it allows each test to be run individually if needed: 

Screen Shot 2022-02-05 at 8 25 00 PM

Note: the changes to `camera_and_viewport_test.dart` are purely cosmetic, even though GitHub shows a large diff.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- (NA) I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

WIP for #1351

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by luanpotter — COMMENTED** [review]
I remember that I annotated the `flameGame.test` so it would be recognized as a test function, at least on vscode. Interesting to see it didn't work, I wonder if we can do something to create custom test functions.

Regardless of that, I don't see any reason not to change the api slightly to your suggestion, it looks equally clean to me. I imagine the goal is to replace the old one entirely right?

I leave it for @erickzanardo to ponder as he originally designed this particular API

**Comment by st-pasha** [comment]
Yes, I was wondering about that too...
My suspicion is that `@isTest` only works with pure functions, not with methods.

So I can actually change this current function into `@isTest`-declared `flameGameTest()`, in which case it would be much closer to the original, while still maintaining the ability to be run individually:

Screen Shot 2022-02-09 at 8 43 11 AM

What do you think?

**Review by spydon — APPROVED** [review]

**Comment by st-pasha** [comment]
@spydon Do you think it's better to have
```dart
test(
  'test name',
  withFlameGame((game) async {
    ...
  }),
);
```
or
```dart
testFlameGame(
  'test name',
  (game) async {
    ...
  },
);
```
?

**Comment by spydon** [comment]
> @spydon Do you think it's better to have
> ```dart
> test(
>   'test name',
>   withFlameGame((game) async {
>     ...
>   }),
> );
> ```
> or
> ```dart
> testFlameGame(
>   'test name',
>   (game) async {
>     ...
>   },
> );
> ```
> ?

Not sure, maybe the first? 🤔
Is there any standard for how others are doing this?

**Comment by luanpotter** [comment]
@st-pasha I see, that must be the case, good catch. mystery solved aside, as for which is better, I don't have any strong preference. maybe @erickzanardo can chime in as he designed the original api (before my refactor)

**Comment by st-pasha** [comment]
> I do prefer this: `testFlameGame( ...`

:+1:
I'll switch the API then.

File: packages/flame_test/lib/src/test_flame_game.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+///     expect(component.isMounted, true);
+///   },
+/// );
+/// ```
+///
+/// The `game` instance supplied by this function to your test [testBody] is a
+/// standard [FlameGame]. If you want to have any other game instance, use the
+/// [testUserGame] function.
+@isTest
+Future<void> testFlameGame(
```
Comment (erickzanardo) [inline packages/flame_test/lib/src/test_flame_game.dart:27]: We should add the equivalent of these methods to the game widget test method, like we have on the Flame Tester class, wdyt?

Comment (st-pasha) [reply packages/flame_test/lib/src/test_flame_game.dart:27]: Probably, though I'm not very familiar with how the widget tests work.

Comment (erickzanardo) [reply packages/flame_test/lib/src/test_flame_game.dart:27]: you could base on the widgetTest on the FlameTester class.

But we could also leave that to a follow up too

File: packages/flame_test/pubspec.yaml [hunk 2]
```diff
@@ -14,7 +14,7 @@ dependencies:
   flutter_test:
     sdk: flutter
   meta: ^1.7.0
-  test: ^1.19.3
+  test: ^1.17.12
```
Comment (erickzanardo) [inline packages/flame_test/pubspec.yaml:17]: Hmm, not sure about downgrading this, any reason for that?

Comment (erickzanardo) [reply packages/flame_test/pubspec.yaml:17]: Ahhh, got it, this is conflicting with flutter's sdk test package.

We can revisit this later

**Review by erickzanardo — APPROVED** [review]

**Pull request merged**